### ISSUE TITLE
Compatibility with other models (*.pb)? Also, I found a way to use cudnn with this.

### ISSUE THREAD

**Issue by Klanly** [nolan-dev/GANInterface#4]
a. Is this compatible with 1024x1024 models like these:
https://github.com/a312863063/generators-with-stylegan2
?
(It says it's StyleGAN2 though)

b. Are there hardcoded parts in the (C# and the dll) codes?

c. I saved a *.pkl model to *.pb on GoogleColab with this script:

############
import pickle
import dnnlib.tflib as tflib
import tensorflow as tf

EXPORT_PATH = "savemodel2"

tflib.init_tf()
_G, _D, Gs = pickle.load(open("xxx.pkl", "rb"))

sess = tf.get_default_session()

tf.saved_model.simple_save(
    sess,
    EXPORT_PATH,
    inputs={t.name:t for t in Gs.input_templates},
    outputs={t.name:t for t in Gs.output_templates})
############

(from http://tachibana.ai/2019/06/15/Portal-Log-One.html)

It generates a pb and a "Variables" folder (Weights not embed inside pb?), Can this kind of saved binary be loaded?
  

  

  
=============   

Also I found that replacing tensorflow.dll with Topaz ones from
https://d2xkriaa67cpt4.cloudfront.net/packages/win/topazmask/com.topazlabs.topazmask.libs3rdparty/1.0.7topazmasklibs3rdparty.7z
can make this use cudnn:

with cudnn:
![cudnn](https://user-images.githubusercontent.com/11384098/71834941-5f261300-30eb-11ea-9f9d-8c006ed74b2c.jpg)

without cudnn:
![nocudnn](https://user-images.githubusercontent.com/11384098/71834959-69e0a800-30eb-11ea-9c4e-4f6e77f22ddf.jpg)

Any of you who has interest can try yourself. This one
https://d2xkriaa67cpt4.cloudfront.net/packages/win/topazstudio2/com.topazlabs.topazstudio2.libs3rdparty/2.1.1topazstudio2libs3rdparty.7z
(TF 1.14, no cudnn) works too.

**Comment by nolan-dev** [comment]
Hi Klanly,

I've been intending to provide an update to work with the official implementation of stylegan (and stylegan2), but I've been distracted by another project.   It'll require a couple minor modifications of the official stylegan2 graph in order to allow modifying the feature maps, which makes the process a bit more complicated than the script you mentioned.  Now the best way I can think of to do that is to load the pkl, save the weights, then load the weights with a slightly modified version of the official implementation, then save/freeze the final graph.  I did an update a little while ago to specify the names of the tensors in the graph that get read from/written to in https://github.com/nolan-dev/GANInterface/blob/master/GanStudio/GanStudio/example_config.xml rather than hardcoding them, so the tool shouldn't need much modification to work with the styelgan2 graph, but the input tensors (like generator_1/Res4/FmapRes4) correspond to operations that aren't present in the official implementation by default.

My plan is to do another blog with instructions/code for converting a stylegan/stylegan2 pkl file, along with a significant update to the tool targeted at automating the detection and modification of features (that part's done, but in a private repo because I want to do a few more bug checks).  

In summary:

a. Is this compatible with 1024x1024 models like these:
Not yet, but I hope to fix that within a month

b. Are there hardcoded parts in the (C# and the dll) codes?
I believe there are in the compiled version of the tool I uploaded, but in the current code there shouldn't be (though the config xml will need to be modified).  When I do the blog I'll provide a compiled version of the latest code.

c. I saved a *.pkl model to *.pb on GoogleColab with this script: 
I really wish this was all that was needed but it's going to be a bit more complicated.

Thanks for the tip about the Topaz dll, that looks like a great speedup.

**Comment by Klanly** [comment]
I append a EncodePng node to some existing *.pkl models and freeze them into *.pb
classic Portrait:
https://drive.google.com/file/d/11q-41kW1ATOkJLTu-1pPCE0ZSdFVg118/view

anime (2019-04-30-stylegan-danbooru2018-portraits-02095-066083.pkl):
https://drive.google.com/file/d/116NVi11rsXwg_8uiUprNnj5HdtphW2Ns/view

It can be evaluated by script
```python
import tensorflow as tf
import numpy as np



tf.reset_default_graph()
sess = tf.Session()
with open("output_model2/arty.pb", 'rb') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')




sess.run(tf.global_variables_initializer())
lat = sess.graph.get_tensor_by_name('Gs/latents_in:0')
otpng = sess.graph.get_tensor_by_name('output:0')
latents0 = np.random.randn(1,512).astype(np.float32)
ret = sess.run(otpng,  feed_dict={lat: latents0})
with open('testapng.png', mode='wb') as f:
        f.write(ret)
```
Don't know about the c++ side.

There's also a frozen stylegan2 model:
https://drive.google.com/file/d/12EtY4ZiuuKyU41E_TPcTzyN65FHv7zPY/view

and fused_bias_act, upfirdn_2d dlls which required for c++
[stylegan2ops.zip](https://github.com/nolan-dev/GANInterface/files/4072486/stylegan2ops.zip)

I guess these op dlls should be loaded with TF_Library* lib = TF_LoadLibrary(xxx) ---> ??? = TF_GetOpList(lib),
but can't find any usage examples (in c/c++ apis) on internet.

**Comment by nolan-dev** [comment]
Thanks for the input, it has been very helpful.  Finally managed to pull myself away from the other stuff I've been working on, and I think I'm close to getting everything working with stylegan2 models https://i.imgur.com/Whmjrsl.png.  I'm using the approach I mentioned before: saving the weights stored in the pkl, then loading them into a modified version of the graph.  That means I can use the reference implementation of the custom ops, which simplifies things.  The approach is very hackish, but with a little effort I think the method will work for most official stylegan2 models.  There's still some stuff I need to clean up, but I'll try and get the updates uploaded soon.

**Comment by nolan-dev** [comment]
Sorry for the delay.  Still not totally finished, but here's an example of the next version of the tool with a frozen ffhq graph: https://drive.google.com/file/d/13JlhTn3FUZJo0822eBs0_EOXoGc8mr7v/view?usp=sharing
This is a compiled version of the code that is (as of this writing) in the ui_and_scripting_update branch: https://github.com/nolan-dev/GANInterface/tree/ui_and_scripting_update.  I'll merge that after testing it a bit more.

The method I have for converting pkl to pb is still really hacky, to the point where I had it working last week, got sick and didn't work on it, and now I'm having to rediscover how I did it all over again lol.  I'll try and clean that up and share it soon.

**Comment by nolan-dev** [comment]
Added instructions/tools for converting from pkl to graph here: https://github.com/nolan-dev/GANInterface/tree/master/pkl_conversion
